### PR TITLE
[SPARK-6408] [SQL] Fix JDBCRDD filtering string literals

### PR DESCRIPTION
**Pull request by ypcat**

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala [hunk 1]
```diff
@@ -129,6 +129,9 @@ class JDBCSuite extends FunSuite with BeforeAndAfter {
     assert(sql("SELECT * FROM foobar WHERE THEID < 1").collect().size == 0)
     assert(sql("SELECT * FROM foobar WHERE THEID != 2").collect().size == 2)
     assert(sql("SELECT * FROM foobar WHERE THEID = 1").collect().size == 1)
+    assert(sql("SELECT * FROM foobar WHERE NAME = 'fred'").collect().size == 1)
+    assert(sql("SELECT * FROM foobar WHERE NAME > 'fred'").collect().size == 2)
+    assert(sql("SELECT * FROM foobar WHERE NAME != 'fred'").collect().size == 2)
```
Comment (liancheng) [inline sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala]: Could you please also add test cases for string literals which contain escaped single / double quotes?

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala [hunk 2]
```diff
@@ -230,13 +231,19 @@ private[sql] class JDBCRDD(
...
+    def compileValue(value: Any): Any = value match {
+      case stringValue: String => s"'${escapeSql(stringValue)}'"
+      case _ => value
+    }
+    f match {
+      case EqualTo(attr, value)            => s"$attr = ${compileValue(value)}"
+      case LessThan(attr, value)           => s"$attr < ${compileValue(value)}"
+      case GreaterThan(attr, value)        => s"$attr > ${compileValue(value)}"
+      case LessThanOrEqual(attr, value)    => s"$attr <= ${compileValue(value)}"
+      case GreaterThanOrEqual(attr, value) => s"$attr >= ${compileValue(value)}"
```
Comment (liancheng) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala]: Please don't align the code vertically, this makes future changes more difficult.

**Comment by ypcat** [comment]
@liancheng is it better to create another table for escape test case? I see no existing test tables suitable for this and it might break existing test cases.

**Comment by liancheng** [comment]
Can we update a single row in the `people` table? For example, change "joe" to "joe armstrong". Then I think only the "SELECT first field" case needs to be updated.

**Comment by ypcat** [comment]
ok, but I presume we are going to test single and double quotes, so it would be like 'joe' "armstrong"?

**Comment by liancheng** [comment]
Yeah, sure :)

**Comment by ypcat** [comment]
Updated as discussed. Tested on my laptop.

**Comment by liancheng** [comment]
ok to test

**Comment by ypcat** [comment]
oops, actually I seem not to really test it.

**Comment by ypcat** [comment]
I think I found another bug:
SQLContext cannot handle quote escaping as HiveContext does, for example:
sql("select '\\'foo\\' \"bar\"'")
I'll investigate further and file another bug

**Comment by ypcat** [comment]
unit test should pass now.

File: sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala [hunk 3]
```diff
@@ -230,13 +231,19 @@ private[sql] class JDBCRDD(
...
    */
-  private def compileFilter(f: Filter): String = f match {
-    case EqualTo(attr, value) => s"$attr = $value"
-    case LessThan(attr, value) => s"$attr < $value"
-    case GreaterThan(attr, value) => s"$attr > $value"
-    case LessThanOrEqual(attr, value) => s"$attr <= $value"
-    case GreaterThanOrEqual(attr, value) => s"$attr >= $value"
-    case _ => null
+  private def compileFilter(f: Filter): String = {
+    def compileValue(value: Any): Any = value match {
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/jdbc/JDBCRDD.scala]: I would avoid nested functions unless they actually need to be nested (i.e. they are using something that is only available in the scope of the function).

**Comment by liancheng** [comment]
Thanks! Merged into master and branch-1.3.

**Pull request closed without merging**